### ISSUE TITLE
Malformed FAT validation error

### ISSUE THREAD

**Issue by jtran** [mdsteele/rust-cfb#8]
I'm getting this error on one of my files when I try to open it for reading:

```
Malformed FAT (FAT has 896 entries, but file has only 811 sectors)
```

But I've used this file with other programs, and they're able to use it just fine.

Is it possible that there's a problem with the validation logic?

I can either post or send you the problem file.  How would you like me to send it to you?  I don't think GitHub will let me attach it.

**Comment by mdsteele** [comment]
If the file has 811 sectors, it sounds like it's no more than a few MB, so it ought to be small enough to attach here (https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/attaching-files).  You might need to put it into a .zip file for GitHub to accept it?

**Comment by jtran** [comment]
Ah, zipping it worked.  Thanks.  [Fans-base.SchDoc.zip](https://github.com/mdsteele/rust-cfb/files/7071013/Fans-base.SchDoc.zip)

**Comment by mdsteele** [comment]
Thanks, I think I've figured out the problem.

**Short version:** This file is technically violating the CFB spec, but since other programs accept it, we should make the `cfb` crate accept it too.  I'll look into it.

**Longer version:** Each FAT sector in a v3 CFB file contains allocation entries for exactly 128 sectors.  If the number of sectors in the file isn't a multiple of 128, then the last FAT sector will have extra entries for beyond the end of the file.  According to the [MS-CFB spec](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-cfb/53989ce4-7b05-4f8d-829b-d08d6148375b), section 2.3:

> The last FAT sector can have more entries that span past the actual size of the compound file. In this case, the entries that cover past end-of-file MUST be marked with FREESECT (0xFFFFFFFF). The size of a compound file is determined by the index of the last non-free FAT array entry. If the last FAT sector contains an entry FAT[N] != FREESECT (0xFFFFFFFF), the file size MUST be at least (N + 1) x (Sector Size) bytes in length.

However, in this file, the extra FAT entries are all initialized to zero, rather than FREESECT.  Technically, that means that the file is in violation of the spec, so the `cfb` crate's validation logic is correct to complain.  But probably the `cfb` crate is being unnecessarily strict here.  (This isn't the time that's happened; it turns out that lots of CFB files in the wild violate the spec in various ways.)  So I'll look into changing this part of the validation logic.

**Comment by jtran** [comment]
Thanks so much for looking into this!  I'm not too familiar with the CFB file format, but I'm trying to learn.  In addition to the documentation you linked in the readme, I've been reading the source of CFB/OLE libraries in other languages to try to understand what's going on.  I've also used this python library's [1] CLI to see the FAT.  If there's something simple I can help with, mention me, and I may be able to help with a pull request.  I'd love to be able to use Rust for what I'm working on.

1: https://github.com/decalage2/olefile

**Comment by mdsteele** [comment]
I think this should be fixed now; with commit bd8600d, I'm now able to open that file.
